app/uk/gov/hmrc/transitmovements/controllers/MessageBodyController.scala

```diff
@@ -94,7 +96,16 @@ class MessageBodyController @Inject() (
           messageType   <- extract(request.headers).asPresentation
           messageData   <- messagesXmlParsingService.extractMessageData(request.body, messageType).asPresentation
           extractedData <- movementsXmlParsingService.extractData(messageType, request.body).asPresentation
-          bodyStorage   <- messageService.storeIfLarge(movementId, messageId, size, request.body).asPresentation
+          _ <- messageType match {
+            case MessageType.DeclarationData =>
+              repo
+                .restrictDuplicateLRN(
+                  extractedData.flatMap(_.localReferenceNumber).get
+                )
+                .asPresentation
+            case _ => EitherT.rightT[Future, MongoError]((): Unit).asPresentation
```

Comment:
Not convinced we should do this right now, but I did want to get my thoughts down to possibly sort this later as part of the object-store rationalisation stuff.

---

At some point I guess I'm splitting hairs, but the LRN is only extracted for declaration data anyway, and for neatness it would be better to keep pattern matching out of this comprehension, which is a series of steps that is easy to read.

Perhaps create a method like so (implicits may need sorting out too):

```scala
def checkForDuplicateLRN(movement: Option[Movement]): EitherT[Future, MongoError, Unit] = 
    message match {
        case Some(Movement(_, _, _, _, _, Some(lrn), _, _, _)) => repo.restrictDuplicateLRN(lrn)
        case None => EitherT.rightT[Future, MongoError]((): Unit)
    }
``` 

then just use this method in the for comprehension.

We have to do this in the `MovementsController` too and this should really be part of a service, hence why I'm not going to insist on this right now

Reply:
Thank you @drnaylor-oc for the suggestion, as it might get moved after the object-store rationalisation I better leave it here.